File: src/audioWfc/wfc/hierarchy/SectionLevelNode.java
package audioWfc.wfc.hierarchy;

import audioWfc.musicTheory.OctavedNote;
import audioWfc.wfc.CanvasAttributes;
import audioWfc.wfc.HigherValues;
import audioWfc.wfc.TileCanvas;
import audioWfc.wfc.hierarchy.prototypes.ChordPrototype;
import audioWfc.wfc.hierarchy.prototypes.Chordesque;
import audioWfc.wfc.hierarchy.prototypes.Section;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class SectionLevelNode {
    private HigherValues higherValues;
    private CanvasAttributes<Chordesque> chordAttributes;
    private CanvasAttributes<OctavedNote> noteAttributes;
    private TileCanvas<Section> canvas;
    private Random random;

    public SectionLevelNode(HigherValues higherValues, CanvasAttributes<Section> sectionAttributes, CanvasAttributes<Chordesque> chordAttributes, CanvasAttributes<OctavedNote> noteAttributes, Random random){
        this.higherValues = higherValues;
        this.chordAttributes = chordAttributes;
        this.noteAttributes = noteAttributes;
        this.canvas = new TileCanvas<>(
                sectionAttributes.getSize(),
                sectionAttributes.getOptions(),
                sectionAttributes.getConstraints(),
                higherValues,
                random);
        this.random = random;
    }

    public List<SectionResult> generate(){
        List<Section> sections = canvas.generate();
        List<SectionResult> out = new ArrayList<>();
        for (Section section : sections) {
            ChordLevelNode chordLevelNode = new ChordLevelNode(this, higherValues.copyWithSection(section), chordAttributes.union(section.getChordAttributes()), noteAttributes.union(section.getNoteAttributes()), random);
            List<ChordResult> chords = chordLevelNode.generate();
            out.add(new SectionResult(chords));
        }
        return out;
    }
}

File: vite-project/src/wfc/hierarchy/SectionLevelNode.ts
import { OctavedNote } from "../../music_theory/Note"
import { RhythmPattern } from "../../music_theory/Rhythm"
import { Random } from "../../util/Random"
import { HigherValues } from "../HigherValues"
import { TileCanvasProps, TileCanvas } from "../TileCanvas"
import { ChordLevelNode } from "./ChordLevelNode"
import { Chordesque, Section } from "./prototypes"
import { SectionResult, SectionResultWithRhythm } from "./results"

interface SectionLevelNodeProps {
  higherValues?: HigherValues;
  noteCanvasProps: TileCanvasProps<OctavedNote>;
  chordesqueCanvasProps: TileCanvasProps<Chordesque>;
  rhythmPatternCanvasProps: TileCanvasProps<RhythmPattern>;
  sectionCanvasProps: TileCanvasProps<Section>;
  random: Random;
}

export class SectionLevelNode {
	private higherValues: HigherValues
	private noteCanvasProps: TileCanvasProps<OctavedNote>
	private chordesqueCanvasProps: TileCanvasProps<Chordesque>
	private rhythmPatternCanvasProps: TileCanvasProps<RhythmPattern>
	private sectionCanvas: TileCanvas<Section>
	private random: Random

	constructor(props: SectionLevelNodeProps){
		this.higherValues = props.higherValues ?? new HigherValues()
		this.noteCanvasProps = props.noteCanvasProps
		this.chordesqueCanvasProps = props.chordesqueCanvasProps
		this.rhythmPatternCanvasProps = props.rhythmPatternCanvasProps
		this.sectionCanvas = new TileCanvas(props.sectionCanvasProps, this.higherValues, props.random)
		this.random = props.random
	}

	public generateWithoutRhythm() : SectionResult[] {
		const sections = this.sectionCanvas.generate()
		return sections.map(section => {
			const chordLevelNode = new ChordLevelNode({
				higherValues: this.higherValues.copyWithSection(section),
				noteCanvasProps: this.noteCanvasProps.union(section.getNoteCanvasProps()),
				chordesqueCanvasProps: this.chordesqueCanvasProps.union(section.getChordesqueCanvasProps()), 
				rhythmPatternCanvasProps: this.rhythmPatternCanvasProps.union(section.getRhythmPatternCanvasProps()), 
				random: this.random
			})
			return chordLevelNode.generateWithoutRhythm()
		})
	}

	public generateWithRhythm() : SectionResultWithRhythm[] {
		const sections = this.sectionCanvas.generate()
		return sections.map(section => {
			const chordLevelNode = new ChordLevelNode({
				higherValues: this.higherValues.copyWithSection(section), 
				noteCanvasProps: this.noteCanvasProps.union(section.getNoteCanvasProps()), 
				chordesqueCanvasProps: this.chordesqueCanvasProps.union(section.getChordesqueCanvasProps()), 
				rhythmPatternCanvasProps: this.rhythmPatternCanvasProps.union(section.getRhythmPatternCanvasProps()), 
				random: this.random
			})
			return chordLevelNode.generateWithRhythm()
		})
	}
}